src/use-swr.ts

```diff
@@ -235,70 +237,56 @@ export const useSWRHandler = <Data = any, Error = any>(
             // case 3
             MUTATION_END_TS[key] === 0)
         ) {
-          setState(newState)
+          finishRequestAndUpdateState()
           return false
         }
 
-        if (!isUndefined(stateRef.current.error)) {
-          newState.error = UNDEFINED
-        }
-
         // Deep compare with latest state to avoid extra re-renders.
         // For local state, compare and assign.
         if (!compare(stateRef.current.data, newData)) {
           newState.data = newData
         }
+
         // For global state, it's possible that the key has changed.
         // https://github.com/vercel/swr/pull/1058
         if (!compare(cache.get(key), newData)) {
           cache.set(key, newData)
         }
-
-        // merge the new state
-        setState(newState)
-
-        if (shouldStartNewRequest) {
-          // also update other hooks
-          broadcastState(cache, key, newData, newState.error, false)
-        }
       } catch (err) {
         cleanupState()
-        cache.set(keyValidating, false)
-        if (getConfig().isPaused()) {
-          setState({
-            isValidating: false
-          })
-          return false
-        }
 
-        // Get a new error, don't use deep comparison for errors.
-        cache.set(keyErr, err)
-        if (stateRef.current.error !== err) {
-          // Keep the stale data but update error.
-          setState({
-            isValidating: false,
-            error: err as Error
-          })
-          if (shouldStartNewRequest) {
-            // Broadcast to update the states of other hooks.
-            broadcastState(cache, key, UNDEFINED, err, false)
-          }
-        }
+        // Not paused, we continue handling the error. Otherwise discard it.
+        if (!getConfig().isPaused()) {
+          // Get a new error, don't use deep comparison for errors.
+          cache.set(keyErr, err)
+          newState.error = err as Error
 
-        // Error event and retry logic.
-        if (isCallbackSafe()) {
-          getConfig().onError(err, key, config)
-          if (config.shouldRetryOnError) {
-            // When retrying, dedupe is always enabled
-            getConfig().onErrorRetry(err, key, config, revalidate, {
-              retryCount: (opts.retryCount || 0) + 1,
-              dedupe: true
-            })
+          // Error event and retry logic.
+          if (isCallbackSafe()) {
+            getConfig().onError(err, key, config)
+            if (config.shouldRetryOnError) {
+              // When retrying, dedupe is always enabled
+              getConfig().onErrorRetry(err, key, config, revalidate, {
```

Comment:
the retry is now before state update, will this lead to any inconsistence?

Reply:
Should be good because state update will be scheduled til next tick, and this was called synchronously.